### PR TITLE
Laplacian regularization

### PR DESCRIPTION
**Pull request by mperego**
This PR contains (each point is a separate commit):

- a fix to the  `L2ProjectedBoundaryLaplacianResidual` which was buggy
- an update of the Humboldt mesh and relative tests. The new mesh is a bit smaller and from a more recent dataset
- a new response, `Quadratic Linear Operator Based`, that computes `p A' inv(D) A p`, where p is a field and `A` and `D` are matrices read from file, `D` being diagonal
- added/modified tests to use the response for computing the Laplacian squared regularization, where `A` is a stiffening matrix (including a mass and robin boundary condition term) and `D` is a lumped mass matrix. The response results are equivalent to those computed with the` Boundary Squared L2 Norm` response together with the `L2ProjectedBoundaryLaplacianResidual` evaluator

### CODE REVIEW

File: src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    loadLinearOperator();
+
+  Teuchos::RCP<const Thyra_Vector> field = app_->getDistributedParameterLibrary()->get(field_name_)->vector();
+
+  // A p
+  A_->apply(Thyra::EOpTransp::NOTRANS, *field, vec1_.ptr(), 1.0, 0.0);
+
+  // coeff inv(D) A p
+  vec2_->assign(0.0);
+  Thyra::ele_wise_divide( coeff_, *vec1_, *D_, vec2_.ptr() );
```
Comment (bartgol) [inline src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp:93]: Thyra::VectorBase has a method to compute a weighted 2 norm. You could just do (here and in the other evaluate methods)
```
g->assign (std::pow(vec1->norm_2(*D_),2));
```
and do everything in one loop.

Comment (bartgol) [reply src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp:93]: NVM, the Tpetra adapter impl does precisely what you are doing.

File: tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch_fluxdiv.yaml [hunk 2]
```diff
@@ -105,7 +115,7 @@ ANONYMOUS:
     Exodus Output File Name: output/humboldt.exo
     Element Shape: Wedge
     Use Serial Mesh: false
-    Columnwise Ordering: true
+    Columnwise Ordering: false
```
Comment (jewatkins) [inline tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch_fluxdiv.yaml:118]: What's the reasoning behind this change?

**Review by bartgol — APPROVED** [review]
Looks good.

**Review by jewatkins — APPROVED** [review]
Looks good.

File: tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch_fluxdiv.yaml (hunk 2 above)
Comment (mperego) [reply tests/small/LandIce/FO_GIS/input_fo_humboldt_frosch_fluxdiv.yaml:118]: The new response works only for contiguous maps (it would error out otherwise). When the columnwise ordering is true, the basal map is not contiguous.

File: src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp (hunk 1 above)
Comment (mperego) [reply src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp:93]: here I need to divide by `D`, not multiply by `D`.

**Comment by mperego** [comment]
Thanks @bartgol and @jewatkins for the review!

File: src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp (hunk 1 above)
Comment (bartgol) [reply src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp:93]: Ah, right.
